### ISSUE TITLE
`pnpm deploy` doesn't rewrite `workspace:*` dependencies in `package.json`

### ISSUE THREAD

**Issue by rijkvanzanten** [pnpm/pnpm#6269]
### pnpm version: 

`7.30.0`

### Code to reproduce the issue: 

Run `pnpm -F <project> deploy`

### Expected behavior: 

When the target directory is created, I expected the `package.json` to be "updated" to include the workspace packages with their respective version numbers similar to `pnpm pack` or `pnpm publish`. 

### Actual behavior: 

`workspace:*` versions remain in package.json. This in turn means consumers of the output directory can't rely on npm/yarn or any other package manager. A temporary workaround would be to `pnpm pack` the target directory, then pull the "clean" `package.json` from the tarball, and put that back into the target directory of the `deploy`. 

### Additional information:

 - `node -v` prints: v18.12.1
 - Windows, macOS, or Linux?: any

**Comment by pkerschbaum** [comment]
Had the same problem, and solved it with a script which replaces the `workspace:` protocol of workspace dependencies by the respective version number.  
Maybe it helps you as a workaround.

```js
// replace-workspace-protocol-by-version.mjs
import fs from 'node:fs';
import path from 'node:path';
import url from 'node:url';
import { Command } from '@commander-js/extra-typings';
import { findWorkspacePackagesNoCheck } from '@pnpm/find-workspace-packages';

const program = new Command().argument(
	'<pathtopackagejson>',
	'the path to the package.json of which the dependencies using the "workspace:*" protocol should be replaced by the versions present in the monorepo (e.g. "./deployed-package/package.json")'
);

program.parse();

const __dirname = url.fileURLToPath(new URL('.', import.meta.url));

const projects = await findWorkspacePackagesNoCheck(path.join(__dirname, '..', '..', '..'));
// first elem of returned array is the root package.json, the monorepo itself --> discard that
const [_rootProject, ...workspaceProjects] = projects;

const absolutePath = path.join(process.cwd(), program.args[0]);

const packageJsonContent = JSON.parse(await fs.promises.readFile(absolutePath, 'utf-8'));

/**
 * @param {{[key: string]: string}} dependencies
 */
function replaceWorkspaceProtocolBySpecificVersion(dependencies) {
	if (typeof dependencies !== 'object' || dependencies === null) {
		return;
	}

	for (const [pkgName, pkgVersion] of Object.entries(dependencies)) {
		if (pkgVersion.startsWith('workspace:')) {
			const matchingWorkspaceProject = workspaceProjects.find(
				project => project.manifest.name === pkgName
			);
			if (!matchingWorkspaceProject) {
				throw new Error(
					`could not replace the workspace protocol of a package, reason: package not found! pkgName=${pkgName}`
				);
			}
			if (!matchingWorkspaceProject.manifest.version) {
				throw new Error(
					`could not replace the workspace protocol of a package, reason: package has no version! pkgName=${pkgName}`
				);
			}

			dependencies[pkgName] = `^${matchingWorkspaceProject.manifest.version}`;
		}
	}
}

replaceWorkspaceProtocolBySpecificVersion(packageJsonContent.dependencies);
replaceWorkspaceProtocolBySpecificVersion(packageJsonContent.devDependencies);
replaceWorkspaceProtocolBySpecificVersion(packageJsonContent.peerDependencies);
replaceWorkspaceProtocolBySpecificVersion(packageJsonContent.optionalDependencies);

await fs.promises.writeFile(
	absolutePath,
	JSON.stringify(packageJsonContent, undefined, 2),
	'utf-8'
);
```

```sh
# run it against the package.json of the deployed package
node ./replace-workspace-protocol-by-version.mjs ./deployed-package/package.json
```

**Comment by alvis** [comment]
It appears that it's a careless mistake which still remains in version 8.0.0-rc1. Nevertheless, it is not a significant issue since the code is capable of running without the need for package.json.

**Comment by alvis** [comment]
> This really depends on your actual use case. In my case, it's a major blocker that requires the workaround 🤷🏻

@rijkvanzanten Just curious. If you need to read the version of your dependencies at runtime, you probably want to get it like
```ts
import libPackageJson from 'lib_name/package.json';
const libVersion = libPackageJson.version;
```

Alternatively you can get the versions from the lock file.

**Comment by alvis** [comment]
@rijkvanzanten Right, I guess I got you. Like using `read-pkg-up` which would try to `normalize` the content of `package.json` and as it's not a valid one it throw an error?

**Comment by rijkvanzanten** [comment]
I'm distributing an app through a Docker image that relies on `pnpm deploy` to generate the final "dist" that's shipped. To keep the image lightweight, it's "just" node18-alpine and our dist folder, that in turn also means that the user may or may not have to interact with the packages through `npm` instead of `pnpm`. Even though it's recommended to enable pnpm with corepack in that case, that's not always followed or feasible. When trying to use any `npm` command against the package (like `npx` to run the `bin`), it'll fail as the package.json file is "invalid" in that context 👍🏻

**Comment by JacobLey** [comment]
I believe this is addressed by https://github.com/pnpm/pnpm/pull/6943

Which is not released yet, but hopefully will be soon

**Comment by rijkvanzanten** [comment]
Perfect! Thanks Jacob 🙂

**Comment by dgellow** [comment]
Hmm, unfortunately #6943 has been reverted by https://github.com/pnpm/pnpm/pull/7058.

**Comment by kael-shipman** [comment]
I'm experiencing something similar which indicates the broader issue is actually that `pnpm deploy` doesn't apply pnpm's special publishing process at all when it exports the workspace deps. In my case, it's also not applying any of the items listed in `publishConfig`, which results in dependencies that won't import anymore because `"main": "src/index.ts"` is not correctly replaced with `"main": "dist/index.js"`. And since this affects imports in the deployed app, it does require a work-around.

See [this tag](https://github.com/wymp/ts-monorepo-poc/releases/tag/pnpm-bug) for a reproducible example. (This is a POC monorepo, so should be fairly minimal.)

In my case, I have several workspace libraries that need to be "published" into the pruned deploy. None of these libraries are correctly processed and so none of them import when the program is executed. @rijkvanzanten's solution is a good start, but unfortunately it doesn't work out of the box for the libraries that are copied into the `pruned` folder because you can't run `pnpm pack` successfully from inside a subfolder of `node_modules` (because it doesn't resolve its own dependencies correctly).

Since my use-case is also docker, I have a little more leeway to do messy or destructive things, so my work-around is somewhat brutish:

```dockerfile
RUN \
  pnpm --filter my-app build && \
  pnpm --filter my-app --prod deploy /pruned/my-app && \
  sed -ri 's#"main": "src/(.+)\.ts"#"main": "dist/\1.js"#' /pruned/my-app/node_modules/@my-namespace/*/package.json
```

**Comment by flevi29** [comment]
It might be somewhat related, in my case since `deploy` doesn't create a flat `node_modules`, the docker image cannot find packages that would be in the workspace root `package.json` `dependencies`. But strangely this only happens in docker, on local deploy seems to work fine, it's probably more nuanced.

Point is all of this added complexity and headache could be entirely avoided if `pnpm deploy` created something `npm` compatible. That would be really nice.

**Comment by forty** [comment]
IMO this is a feature. Some version of PNPM do rewrite the package.json (for example I use 8.7.1 which rewrite the package.json), and this is really anything as this make it impossible to do reproductible build, as writing JSON ends up with random key order.

My use cas is different, I don't use deploy for the repo to be used with npm, but only to create a tarball to be run on a server

**Comment by ricardobeat** [comment]
+1 to this. It is a critical blocker to adopting `pnpm` any organization that has multiple projects on yarn or npm, as simply trying to install local packages using `portal:` or links becomes impossible.

**Comment by y-nk** [comment]
i'm hitting this running 10.13, was this fixed or not?

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm deploy` rewrites a `workspace:*` dependency to a concrete `file:` specifier in the deployed package.json.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
